web3_proxy/src/stats/mod.rs

```diff
@@ -222,6 +223,9 @@ impl BufferedRpcQueryStats {
             .unwrap_or(Decimal::from(0));
     }
 
+    /// Check a user's balance and possibly downgrade him in the cache
+    async fn downgrade_user(self) {}
```

Comment:
No. Don't do it do it like this. See my other comment

Reply:
Also, opening and holding this many locks across multiple users at once seems really likely to me to cause a deadlock.

if I refer you, and we both make queries at the same time on different servers, both of our stats updates will try to lock our users. And your stats update will also try to lock my user. That's a problem. 

I think we need to do the 'Expr' code like I first suggested

Reply:
I changed the logic to use Expr everywhere except on the referral table (because more complex logic). Let me know if this looks ok now. If the lock should still be removed, I think the only way around would be to do an SQL statement (unless sea-orm syntax allows more complex statements). let me know